File: backend/ReadingCommunityApi.Core/Models/Author.cs
namespace ReadingCommunityApi.Core.Models;

public class Author
{
    public int Id { get; private set; }
    public string Name { get; private set; }
    public string Biography { get; private set; }
    public DateTime? BirthDate { get; private set; }
    public DateTime? DeathDate { get; private set; }
    public string Nationality { get; private set; }
    public string ProfileImageUrl { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime? UpdatedAt { get; private set; }
    public ICollection<Book> Books { get; private set; } = new List<Book>();

    private Author() { } //EMpty constructure to the EF Core
    public Author(string name, string biography, DateTime? birthDate,
                  string nationality, string profileImageUrl)
    {
        Name = name;
        Biography = biography;
        BirthDate = birthDate;
        Nationality = nationality;
        ProfileImageUrl = profileImageUrl;
        CreatedAt = DateTime.UtcNow;
    }
    public void UpdateProfileImage(string profileImageUrl)
    {
        ProfileImageUrl = profileImageUrl;
        UpdatedAt = DateTime.UtcNow;
    }

    public void AddBook(Book book)
    {
        Books.Add(book);
        UpdatedAt = DateTime.UtcNow;
    }

    public bool IsAlive => DeathDate == null;
    public int? Age
    {
        get
        {
            if (BirthDate == null) return null;
            var endDate = DeathDate ?? DateTime.UtcNow;
            var age = endDate.Year - BirthDate.Value.Year;
            if (endDate < BirthDate.Value.AddYears(age)) age--;
            return age;
        }
    }   
}

File: frontend/src/types/author.ts
import { BookListDTO } from './book';

// Author DTOs
export interface AuthorListDTO {
  id: number;
  name: string;
  nationality: string;
  numberOfBooks: number;
  profileImageUrl: string;
}

export interface AuthorDetailDTO {
  id: number;
  name: string;
  biography: string;
  birthDate: string | null;
  deathDate: string | null;
  age: number;
  isAlive: boolean;
  nationality: string;
  profileImageUrl: string;
  books: BookListDTO[];
}

export interface AuthorCreateDTO {
  name: string;
  biography: string;
  birthDate: string | null;
  deathDate: string | null;
  nationality: string;
  profileImageUrl: string;
}
